cloud-controller-manager/do/doks_lb_validator.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	return doClient, nil
+}
+
+// DOKSLBServiceValidator ...
+func (v *DOKSLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+	v.Log.V(10).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "bad request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	v.Log.V(10).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return admission.Allowed("")
+	}
+
+	// check if request is for LB deletion
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer && svc.DeletionTimestamp != nil {
+		return admission.Allowed("the loadbalancer is deleting")
+	}
+
+	// initialize DO Client if not already initialized
+	if v.gClient == nil {
+		doClient, err := initDOClient()
+		if err != nil {
+			v.Log.Error(err, "failed to initialize DO client")
+			return admission.Errored(http.StatusConflict, err)
+		}
+		v.gClient = doClient
+	}
+
+	clusterID := os.Getenv(doClusterIDEnv)
+	if clusterID == "" {
+		fmt.Println("missing cluster id")
+	}
+	region := "nyc1"
+	if region == "" {
+		fmt.Println("missing region id")
+	}
+
+	nodePools, _, err := v.gClient.Kubernetes.ListNodePools(ctx, clusterID, &godo.ListOptions{})
+	if err != nil {
+		v.Log.Error(err, "no nodes found")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	var dropletIDs []int
+	for _, pool := range nodePools {
+		for _, node := range pool.Nodes {
+			dropletID, err := strconv.Atoi(node.DropletID)
+			if err != nil {
+				v.Log.Error(err, "failed to retrieve droplet ids")
+				return admission.Errored(http.StatusConflict, err)
+			}
+			dropletIDs = append(dropletIDs, dropletID)
+		}
+	}
+
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest, err := v.buildRequest(svc.Name, region, dropletIDs, forwardingRules)
+	if err != nil {
+		v.Log.Error(err, "failed to build lb request")
+		//fmt.Printf("failed to build lb request", err)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
+		err = v.validateCreate(ctx, lbRequest, v.gClient)
+		if err != nil {
+			v.Log.Error(err, "failed to validate lb creation")
+			return admission.Denied("failed to validate load balancer request")
+		}
+		return admission.Allowed("valid lb create request")
+	}
+
+	err = v.validateUpdate(ctx, svc, lbRequest, v.gClient)
+	if err != nil {
+		fmt.Printf("failed to update load balancer")
+		v.Log.Error(err, "failed to update load balancer")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	return admission.Allowed("validating if request type is UPDATE")
+}
+
+func (v *DOKSLBServiceValidator) validateCreate(ctx context.Context, lbRequest *godo.LoadBalancerRequest, doClient *godo.Client) error {
```

Comment:
If we move `doClient` to `v`, then we don't have to pass it in.

Reply:
In it's current state, `v.gClient` will give you the `doClient` you need. So passing in the `doClient` is redundant.

The only thing needed now, as Timo suggested, is to initialize `v.gClient` in main instead of in this file and also remove the `doClient` value from this function signature and just call `v.gClient` as needed.